src/storage/mutate/AddEdgesProcessor.cpp

```diff
@@ -13,27 +13,163 @@ namespace nebula {
+            /**
+             * skip the in-edge.
+             **/
+            if (index.get_tagOrEdge() == type) {
+                auto ekVal = NebulaKeyUtils::edgeKey(partId, srcId, type, rank, dstId, version);
+                newIndexes[ekVal] = std::make_pair(edge.get_props(), index);
+            }
+        }
     });
+    for (auto& index : newIndexes) {
+        /*
+         * step 1 , Delete old version index if exists.
+         */
+        auto oi = obsoleteIndex(partId, index.first, index.second.second);
+        if (!oi.empty()) {
+            batchHolder->remove(std::move(oi));
+        }
+        /*
+         * step 2 , Insert new edge data
+         */
+        auto key = index.first;
+        auto prop = index.second.first;
+        batchHolder->put(std::move(key), std::move(prop));
+        /*
+         * step 3 , Insert new edge index
+         */
+        auto ni = newIndex(partId, index.first, index.second);
+        batchHolder->put(std::move(ni), "");
+    }
+
+    return encodeBatchValue(batchHolder->getBatch());
+}
+
+std::string AddEdgesProcessor::obsoleteIndex(PartitionID partId,
+                                             const folly::StringPiece& rawKey,
+                                             const nebula::cpp2::IndexItem& index) {
+    auto prefix = NebulaKeyUtils::edgePrefix(partId,
+                                             NebulaKeyUtils::getSrcId(rawKey),
+                                             NebulaKeyUtils::getEdgeType(rawKey),
+                                             NebulaKeyUtils::getRank(rawKey),
+                                             NebulaKeyUtils::getDstId(rawKey));
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kvstore_->prefix(this->spaceId_, partId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "Error! ret = " << static_cast<int32_t>(ret)
+                   << ", spaceId " << this->spaceId_;
+        return "";
+    }
+    if (iter && iter->valid()) {
+        auto reader = RowReader::getEdgePropReader(this->schemaMan_,
+                                                  iter->val(),
+                                                  spaceId_,
+                                                  index.get_tagOrEdge());
+        const auto& cols = index.get_cols();
+        auto values = collectIndexValues(reader.get(), cols);
+        auto indexKey = NebulaKeyUtils::edgeIndexKey(partId,
+                                                     index.get_index_id(),
+                                                     NebulaKeyUtils::getSrcId(rawKey),
+                                                     NebulaKeyUtils::getRank(rawKey),
+                                                     NebulaKeyUtils::getDstId(rawKey),
+                                                     values);
+        std::string val;
+        auto result = kvstore_->get(spaceId_, partId, indexKey, &val);
```

Comment:
What's the purpose about this get.

Reply:
> What's the purpose about this get.

First, we need to check if there is an older version of this edge. If it exists, then index row corresponding to old version data must also exist. So this `get` is redundant. you are right. good point.